src/hotspot/share/classfile/protectionDomainCache.cpp

```diff
@@ -35,23 +35,24 @@
 #include "oops/oop.inline.hpp"
 #include "oops/weakHandle.inline.hpp"
 #include "runtime/atomic.hpp"
+#include "runtime/mutexLocker.hpp"
 #include "utilities/growableArray.hpp"
-#include "utilities/hashtable.inline.hpp"
+#include "utilities/resourceHash.hpp"
 
-unsigned int ProtectionDomainCacheTable::compute_hash(Handle protection_domain) {
-  // Identity hash can safepoint, so keep protection domain in a Handle.
-  return (unsigned int)(protection_domain->identity_hash());
+unsigned int ProtectionDomainCacheTable::compute_hash(const WeakHandle& protection_domain) {
+  return (unsigned int)(protection_domain.resolve()->identity_hash());
 }
 
-int ProtectionDomainCacheTable::index_for(Handle protection_domain) {
-  return hash_to_index(compute_hash(protection_domain));
+bool ProtectionDomainCacheTable::equals(const WeakHandle& protection_domain1, const WeakHandle& protection_domain2) {
+  return protection_domain1.peek() == protection_domain2.peek();
 }
 
-ProtectionDomainCacheTable::ProtectionDomainCacheTable(int table_size)
-  : Hashtable<WeakHandle, mtClass>(table_size, sizeof(ProtectionDomainCacheEntry))
-{   _dead_entries = false;
-    _total_oops_removed = 0;
-}
+ResourceHashtable<WeakHandle, WeakHandle, 1009, ResourceObj::C_HEAP, mtClass,
```

Comment:
I am struggling to understand what the key and values types are in this hashtable ???

Reply:
WeakHandle is the key and the value in this hashtable.  We need to match the WeakHandle in the input by value (see equals function), but we also want to return the WeakHandle to store in to the pd_set list in the DictionaryEntry.